colorizer_data/bin/tfe_open.py

```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python3
+
+from http.server import HTTPServer, SimpleHTTPRequestHandler
+from multiprocessing import Process
+import argparse
+import os
+import signal
+import time
+import webbrowser
+
+# 6465 and 6470 are unassigned ports according to
+# https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml
+default_tfe_port = 6465
+default_directory_port = 6470
+
+
+# Adapted from https://stackoverflow.com/a/21957017.
+class CORSRequestHandler(SimpleHTTPRequestHandler):
+    def end_headers(self):
+        self.send_header("Access-Control-Allow-Origin", "*")
+        SimpleHTTPRequestHandler.end_headers(self)
+
+
+class ReactAppRequestHandler(CORSRequestHandler):
+    def translate_path(self, path):
+        # Serve the React app's index.html for any path that doesn't map to a
+        # file for client-side routing.
+        path = super().translate_path(path)
+        if os.path.exists(path):
+            return path
+        return os.path.join(os.getcwd(), "viewer/index.html")
+
+
+# Adapted from https://stackoverflow.com/questions/18499497/how-to-process-sigterm-signal-gracefully
+class ExitListener:
+    killed = False
+
+    def __init__(self):
+        signal.signal(signal.SIGINT, self.exit_gracefully)
+        signal.signal(signal.SIGTERM, self.exit_gracefully)
+
+    def exit_gracefully(self, signum, frame):
+        print("Exiting...")
+        self.killed = True
+
+
+def is_collection(path):
+    # Can either be a directory containing collection.json or the file itself.
+    if (path.endswith("collection.json")) and os.path.isfile(path):
+        return True
+    return os.path.isdir(path) and os.path.isfile(os.path.join(path, "collection.json"))
+
+
+def is_dataset(path):
+    # Can either be a directory containing manifest.json or the file itself.
+    if (path.endswith("manifest.json")) and os.path.isfile(path):
+        return True
+    return os.path.isdir(path) and os.path.isfile(os.path.join(path, "manifest.json"))
+
+
+def serve_until_terminated(httpd: HTTPServer):
+    # Poll for exit signal every second
+    httpd.timeout = 1
+    httpd.handle_timeout = lambda: None
+    exit_listener = ExitListener()
+
+    while not exit_listener.killed:
+        try:
+            httpd.handle_request()
+        except Exception as e:
+            print("Error occurred while handling request:", e)
+    httpd.server_close()
+
+
+def serve_tfe(port):
+    script_dir = os.path.dirname(os.path.realpath(__file__))
+    os.chdir(script_dir)
+    httpd = HTTPServer(
+        ("localhost", port),
+        ReactAppRequestHandler,
+    )
+    serve_until_terminated(httpd)
```

Comment:
This assumes that the viewer files are in the same directory as the script, under the `viewer` subdirectory.